candidate_relay_test.go

```diff
@@ -22,6 +25,158 @@ func optimisticAuthHandler(string, string, net.Addr) (key []byte, ok bool) {
 	return turn.GenerateAuthKey("username", "pion.ly", "password"), true
 }
 
+func passiveTCPRelayGatherAndExchangeCandidates(agentWithRelay, agentNoRelay *Agent) {
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	check(agentWithRelay.OnCandidate(func(candidate Candidate) {
+		if candidate == nil {
+			wg.Done()
+		}
+	}))
+
+	// This gather will create a single passive TCP relay candidate.
+	check(agentWithRelay.GatherCandidates())
+
+	var tcpCandidateCreated sync.WaitGroup
+	tcpCandidateCreated.Add(1)
+	check(agentNoRelay.OnCandidate(func(candidate Candidate) {
+		if candidate == nil {
+			wg.Done()
+		} else {
+			if candidate.TCPType() != TCPTypeUnspecified {
+				tcpCandidateCreated.Done()
+			}
+		}
+	}))
+
+	// Wait for `agentWithRelay` to gather its candidate. We do not need to gather candidates for
+	// `agentNoRelay`. It will create a candidate in response to receiving a passive TCP candidate.
+	wg.Wait()
+
+	// Communicate the relay candidate to `agentNoRelay`.
+	relayCandidates, err := agentWithRelay.GetLocalCandidates()
+	check(err)
+	for _, relayCandidate := range relayCandidates {
+		candidateCopy, copyErr := relayCandidate.copy()
+		check(copyErr)
+		check(agentNoRelay.AddRemoteCandidate(candidateCopy))
+	}
+
+	// Wait for the TCP candidate to be generated.
+	tcpCandidateCreated.Wait()
+
+	// Assert there's a single tcp candidate. And add it as a remote candidate for `agentWithRelay`.
+	tcpCandidates, err := agentNoRelay.GetLocalCandidates()
+	check(err)
+	if len(tcpCandidates) != 1 {
+		check(fmt.Errorf("Expected 1 TCP candidate. Found: %d", len(tcpCandidates)))
+	}
+
+	for _, c := range tcpCandidates {
+		candidateCopy, copyErr := c.copy()
+		check(copyErr)
+		check(agentWithRelay.AddRemoteCandidate(candidateCopy))
+	}
+}
+
+func passiveTCPRelayConnect(aAgent, bAgent *Agent) (*Conn, *Conn) {
```

Comment:
Same as above -- a copy of `relayConnect`, but using our modified `passiveTCPRelayGatherAndExchangeCandidate` method.